File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        loop: for (int i = 0; i < N; i++) {
            String input = br.readLine();
            int inputSize = input.length();
            Stack<String> stack = new Stack<>();
            for (int j = 0; j < inputSize; j++) {
                //연산시작
                String inputElem = String.valueOf(input.charAt(j));
                if (inputElem.equals("(")) {
                    stack.push(inputElem);
                } else if (inputElem.equals(")")) {
                    if (stack.isEmpty()) {
                        System.out.println("NO");
                        continue loop;
                    } else {
                        stack.pop();
                    }
                }
            }
            if (stack.isEmpty()) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
from sys import stdin as s

#input.txt file
#s = open('input.txt')

n = int(s.readline().rstrip())


for i in range(0, n):
    stack = []
    input = s.readline().rstrip()
    size = len(input)
    vps = True
    for j in range(0, size):
        if(input[j] == "("):
            stack.append("(")
        else:
            if(len(stack)==0):
                vps = False
            else:
                stack.pop()
    if(len(stack) != 0):
        vps = False
    if(vps):
        print("YES")
    else:
        print("NO")

